### PR TITLE
Sparse Tree (#88)

### PR DESCRIPTION
**Pull request by zyfrank**

### CODE REVIEW

File: merkle.go [hunk 1]
```diff
@@ -32,15 +33,56 @@ type Node struct {
 	Right *Node
 }
 
+type HashCountDecorator struct {
```
Comment (lucasvo) [inline merkle.go]: This is only used for testing, right? Could you move it to the tests file?

Comment (zyfrank) [reply merkle.go]: It can be

File: smt.go [hunk 2]
```diff
@@ -0,0 +1,285 @@
+/* Copyright 2013 Steve Leonard <sleonard76@gmail.com>. All rights reserved.
```
Comment (lucasvo) [inline smt.go]: Should probably update that, zyfrank :)

File: smt.go [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+)
+
+// TreeOptions configures tree behavior
+type SMTOptions struct {
+	// EnableHashSorting modifies the tree's hash behavior to sort the hashes before concatenating them
+	// to calculate the parent hash. This removes the capability of proving the position in the tree but
+	// simplifies the proof format by removing the need to specify left/right.
+	EnableHashSorting bool
+
+	// DisableHashLeaves determines whether leaf nodes should be hashed or not. By doing disabling this behavior,
```
Comment (lucasvo) [inline smt.go]:
```suggestion
	// DisableHashLeaves determines whether leaf nodes should be hashed or not. By disabling this behavior,
```

File: smt.go [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+	// you can use a different hash function for leaves or generate a tree that contains already hashed
+	// values.
+	DisableHashLeaves bool
+}
+
+// Tree contains all nodes
+type SMT struct {
+	// All nodes, linear
+	//Nodes []Node
```
Comment (lucasvo) [inline smt.go]:
```suggestion
	// Nodes []Node
```

File: smt.go [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+
+// A Sparse Merkle Tree which support all empty leaves lies in right
+type SMT struct {
+	nodesWithoutEmptyLeavesSubTree [][]Hash
+	leafHashFunc                   hash.Hash
+	nonLeafHashFunc                hash.Hash
+	emptyLeafHash                  []byte
+	levelsUpHashOfEmptyLeaves      []Hash
+	treeHeight                     uint
+	noOfNonEmptyLeaves             int
```
Comment (lucasvo) [inline smt.go]:
```suggestion
	countEmptyLeaves             int
```

File: smt.go [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+	Hash []byte
+	Left bool
+}
+
+// A Sparse Merkle Tree which support all empty leaves lies in right
+type SMT struct {
+	nodesWithoutEmptyLeavesSubTree [][]Hash
+	leafHashFunc                   hash.Hash
+	nonLeafHashFunc                hash.Hash
+	emptyLeafHash                  []byte
```
Comment (lucasvo) [inline smt.go]: I think the code would be simplified a lot if we just invoke leafHashFunc where we need this and remove this.

File: smt.go [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+}
+
+func (self *SMT) RootHash() []byte {
+	if self.noOfNonEmptyLeaves == 0 {
+		return self.levelsUpHashOfEmptyLeaves[len(self.levelsUpHashOfEmptyLeaves)-1]
+	}
+	return self.nodesWithoutEmptyLeavesSubTree[self.treeHeight-1][0]
+}
+
+func (self *SMT) Generate(nonEmptyLeaves [][]byte, totalSize uint64) error {
```
Comment (lucasvo) [inline smt.go]:
```suggestion
func (self *SMT) Generate(leaves [][]byte, totalSize uint64) error {
```

File: smt.go [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+		maxEmtySubTreeHeight++
+	}
+	err := self.computeEmptyLeavesSubTreeHash(maxEmtySubTreeHeight)
+	if err != nil {
+		return err
+	}
+	return self.buildAllLevelNodes(nonEmptyLeaves)
+}
+
+func (self *SMT) GetMerkelProof(leafNo uint) []ProofNode {
```
Comment (lucasvo) [inline smt.go]:
```suggestion
func (self *SMT) GetMerkleProof(leafNo uint) []ProofNode {
```

File: smt.go [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+	Left bool
+}
+
+// A Sparse Merkle Tree which support all empty leaves lies in right
+type SMT struct {
+	nodesWithoutEmptyLeavesSubTree [][]Hash
+	leafHashFunc                   hash.Hash
+	nonLeafHashFunc                hash.Hash
+	emptyLeafHash                  []byte
+	levelsUpHashOfEmptyLeaves      []Hash
```
Comment (lucasvo) [inline smt.go]:
```suggestion
	emptyTreeRootHash      []Hash
```

File: smt.go [hunk 10]
```diff
@@ -0,0 +1,207 @@
...
+type Hash []byte
+
+type ProofNode struct {
+	Hash []byte
+	Left bool
+}
+
+// A Sparse Merkle Tree which support all empty leaves lies in right
+type SMT struct {
+	nodesWithoutEmptyLeavesSubTree [][]Hash
```
Comment (lucasvo) [inline smt.go]:
```suggestion
	fullNodes [][]Hash
```

File: smt.go [hunk 11]
```diff
@@ -0,0 +1,207 @@
...
+	for i := self.treeHeight; i > 1; i-- {
+		err := self.buildInternalOneLevelNodes(i - 1)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (self *SMT) buildLeavesNodes(nonEmptyLeaves [][]byte) error {
```
Comment (lucasvo) [inline smt.go]: I think we shouldn't hash the leaves in the SMT itself, that's a lot less flexible than hashing the leaves before. We should hash the leaves in precise-proofs instead.

File: smt.go [hunk 12]
```diff
@@ -0,0 +1,207 @@
...
+		if err != nil {
+			return err
+		}
+		hashes = append(hashes, hash)
+	}
+	self.nodesWithoutEmptyLeavesSubTree = append(self.nodesWithoutEmptyLeavesSubTree, hashes)
+	return nil
+}
+
+func (self *SMT) buildInternalOneLevelNodes(levelsNo uint) error {
```
Comment (lucasvo) [inline smt.go]:
```suggestion
func (self *SMT) computeNodesAt(level uint) error {
```

File: smt.go [hunk 13]
```diff
@@ -0,0 +1,207 @@
...
+	lastLevelNodesHash := self.nodesWithoutEmptyLeavesSubTree[self.treeHeight-1-levelsNo]
+	counts := len(lastLevelNodesHash)
```
Comment (lucasvo) [inline smt.go]:
```suggestion
	count := len(lastLevelNodesHash)
```

File: smt.go [hunk 14]
```diff
@@ -0,0 +1,207 @@
...
+	hashes := []Hash{}
+	for i := 0; i < counts/2; i++ {
```
Comment (lucasvo) [inline smt.go]: This is nitpicking, but I think it's more elegant to do: 
```suggestion
	for i := 0; i < count; i+=2 {
```

File: smt.go [hunk 15]
```diff
@@ -0,0 +1,207 @@
...
+		hash, err := self.parentHash(lastLevelNodesHash[2*i], lastLevelNodesHash[2*i+1])
```
Comment (lucasvo) [inline smt.go]: With the above change, you can do this:

```suggestion
		hash, err := self.parentHash(lastLevelNodesHash[i], lastLevelNodesHash[i+1])
```

File: smt.go [hunk 16]
```diff
@@ -0,0 +1,207 @@
...
+
+type ProofNode struct {
+	Hash []byte
+	Left bool
+}
+
+// A Sparse Merkle Tree which support all empty leaves lies in right
+type SMT struct {
+	nodesWithoutEmptyLeavesSubTree [][]Hash
+	leafHashFunc                   hash.Hash
```
Comment (lucasvo) [inline smt.go]: As mentioned below, this should be removed and done within precise proofs. It already is here: https://github.com/centrifuge/precise-proofs/blob/master/proofs/tree.go#L556

File: smt.go (hunk 14 above)
Comment (zyfrank) [reply smt.go]: if count is an odd number, then loop number add one, cause failure, take 3 as example

File: smt.go (hunk 16 above)
Comment (zyfrank) [reply smt.go]: nice

File: smt.go (hunk 6 above)
Comment (zyfrank) [reply smt.go]: yes, if leaves provided by precise-proof are all hashes, this make things simpler.

**Pull request merged**